### ISSUE TITLE
FR: local calendars

### ISSUE THREAD

**Issue by knopki** [muness/obsidian-ics#92]
It would be nice to be able to add a local calendar as a file in a vault or as a directory with ics files ([vdir](https://vdirsyncer.pimutils.org/en/stable/vdir.html) format).

Currently, I can't set local file/directory as calendar URL because Obsidian's `request` don't understand this.
Ok, I can get all ics files of one of my calendars like `const icsFiles = app.vault.getFiles().filter(f => f.extension == "ics" && f.path.startsWith("journals/calendars/personal"));`. But I can't reuse `getEvents` because it wants to pull remote calendar. And, alternatively, I can't use `parseIcs`, `extractMeetingInfo`, `filterMatchingEvents` because they are out of default plugin class.

Thanks for the plugin!

**Comment by muness** [comment]
@knopki , interesting, I hadn't heard of vdir. I read the info at the link, and I am not clear on the interface because I don't know how to use it:

- Should the plugin read all the files and simply add each as a calendar, using the `displayname` as the calendar name? Or maybe a new "Add vdir" dialog that allows the same options as the current "Add remote ics" dialog?
- Alternatively, I can expose more of the methods and make them usable from templater/other code.
- What's the easiest way to try out adding a vdir myself to experiment?

**Comment by muness** [comment]
Working on a local env, a la https://github.com/pimutils/vdirsyncer/issues/202#issuecomment-100211119

**Comment by muness** [comment]
Got local calendars working using a vsync of my google calendars using the private ical feeds. I see the structure is one folder per calendar and an ics file per VEVENT.

Will give this a whirl now.

**Comment by knopki** [comment]
@muness for me, exposing the methods is enough.

I think it would be possible, in the add calendar dialog, to offer to specify not the URL, but the path in the repository to the file or directory.
If URL is selected, then the behavior is as it is now.
If a file is selected, then just try to parse that file as a calendar feed.
If a directory is selected, then read all `*.ics`, parse them one by one, add events to the common pile, and then filter by date.
I.e. everything like now, but different functions to access the calendar feed.

Another simple solution. Change only the `getEvent` method: allow passing there an async function that will receive `ICSSettings` calendar as an argument and then return calendar feed.

> I see the structure is one folder per calendar and an ics file per VEVENT.

Yes. And `vdirsyncer` doesn't recourse to the nested directories.

> Will give this a whirl now.

Thank you!

**Comment by muness** [comment]
@knopki , can you give this a try: https://github.com/muness/obsidian-ics/releases/tag/1.5.0

**Comment by knopki** [comment]
@muness I'm testing 1.5.0b

Close button overflow:
![image](https://github.com/muness/obsidian-ics/assets/891832/a3fdbbb3-671d-4d75-800b-ed7f0a790c9e)

- [x] add vdir calendar
- [x] insert from vdir calendar
- [x] use from dataviewjs

Differences with `vdirsyncer`:
- Recursion into directories. Thus, files in subdirectories are also loaded. It's ok.
- You can merge all files into one big feed with multiple `VEVENT`s (`cat *.ics > big.ics`). Or you can download some public feed with all holidays or something like this. It's not ok, it's great.

Also, `main-debug.js` is a nice addition.

To summarize, this release is wonderful. Thank you very much! I'm heading out to tell everyone about this plugin in the local community.

**Comment by muness** [comment]
Thanks so much for the review!
